packages/remix-server-runtime/__tests__/formData-test.ts

```diff
@@ -0,0 +1,175 @@
+    let formData = new NodeFormData();
+    formData.set("a", "value");
+    formData.set("blob", new Blob(["blob".repeat(1000)]), "blob.txt");
+    formData.set("file", new File(["file".repeat(1000)], "file.txt"));
+
+    let req = new NodeRequest("https://test.com", {
+      method: "post",
+      body: formData,
+    });
+
+    let parsedFormData = await parseMultipartFormData(
+      req,
+      async () => undefined
+    );
+
+    expect(parsedFormData.get("a")).toBe(null);
+    expect(parsedFormData.get("blob")).toBe(null);
+    expect(parsedFormData.get("file")).toBe(null);
+  });
+
+  it("can throw errors in upload handlers", async () => {
+    let formData = new NodeFormData();
+    formData.set("blob", new Blob(["blob"]), "blob.txt");
+
+    let req = new NodeRequest("https://test.com", {
+      method: "post",
+      body: formData,
+    });
+
+    let error: Error;
+    try {
+      await parseMultipartFormData(req, async () => {
+        throw new CustomError();
+      });
+      throw new Error("should have thrown");
+    } catch (err) {
+      error = err;
+    }
+    expect(error).toBeInstanceOf(CustomError);
+    expect(error.message).toBe("test error");
+  });
+
+  describe("stream should propagate events", () => {
+    it("when controller errors", async () => {
+      let formData = new NodeFormData();
+      formData.set("a", "value");
+      formData.set("blob", new Blob(["blob".repeat(1000)]), "blob.txt");
+      formData.set("file", new File(["file".repeat(1000)], "file.txt"));
+
+      let underlyingRequest = new NodeRequest("https://test.com", {
+        method: "post",
+        body: formData,
+      });
+      let underlyingBody = await underlyingRequest.text();
+
+      let encoder = new TextEncoder();
+      let body = new ReadableStream({
+        start(controller) {
+          controller.enqueue(
+            encoder.encode(underlyingBody.slice(0, underlyingBody.length / 2))
+          );
+          controller.error(new CustomError());
+        },
+      });
+
+      let req = new NodeRequest("https://test.com", {
+        method: "post",
+        body,
+        headers: underlyingRequest.headers,
+      });
+
+      let error: Error;
+      try {
+        await parseMultipartFormData(req, async () => undefined);
+        throw new Error("should have thrown");
+      } catch (err) {
+        error = err;
+      }
+
+      expect(error).toBeInstanceOf(CustomError);
+      expect(error.message).toBe("test error");
+    });
+
+    it("when controller is closed", async () => {
+      let formData = new NodeFormData();
+      formData.set("a", "value");
+      formData.set("blob", new Blob(["blob".repeat(1000)]), "blob.txt");
+      formData.set("file", new File(["file".repeat(1000)], "file.txt"));
+
+      let underlyingRequest = new NodeRequest("https://test.com", {
+        method: "post",
+        body: formData,
+      });
+      let underlyingBody = await underlyingRequest.text();
```

Comment:
Same question as above. Could possibly simplify this code using `URLSearchParams`.